components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,91 +22,255 @@ import android.os.ParcelFileDescriptor
+    }
+
+    private fun startDownloadJob(download: DownloadState) {
         val notification = try {
             performDownload(download)
-            DownloadNotification.createDownloadCompletedNotification(context, download.fileName)
+            when (downloadJobs[download.id]?.status) {
+                DownloadJobStatus.CANCELLED -> { return }
+
+                DownloadJobStatus.PAUSED -> {
+                    DownloadNotification.createPausedDownloadNotification(context, download)
+                }
+
+                DownloadJobStatus.ACTIVE -> {
+                    DownloadNotification.createDownloadCompletedNotification(context, download)
+                }
+
+                DownloadJobStatus.FAILED -> {
+                    DownloadNotification.createDownloadFailedNotification(context, download)
+                }
+
+                null -> { return }
+            }
         } catch (e: IOException) {
-            DownloadNotification.createDownloadFailedNotification(context, download.fileName)
+            DownloadNotification.createDownloadFailedNotification(context, download)
         }
+
         NotificationManagerCompat.from(context).notify(
+                downloadJobs[download.id]?.foregroundServiceId ?: 0,
+                notification
+        )
+
+        sendDownloadCompleteBroadcast(download.id)
+    }
+
+    private fun registerForUpdates() {
+        val filter = IntentFilter().apply {
+            addAction(ACTION_PAUSE)
+            addAction(ACTION_RESUME)
+            addAction(ACTION_CANCEL)
+            addAction(ACTION_TRY_AGAIN)
+            addAction(ACTION_OPEN)
+        }
+
+        context.registerReceiver(broadcastReceiver, filter)
+    }
+
+    private fun displayOngoingDownloadNotification(download: DownloadState) {
+        val ongoingDownloadNotification = DownloadNotification.createOngoingDownloadNotification(
             context,
-            COMPLETED_DOWNLOAD_NOTIFICATION_TAG,
-            notification
+            download
         )
 
-        val downloadID = intent.getLongExtra(EXTRA_DOWNLOAD_ID, -1)
-        sendDownloadCompleteBroadcast(downloadID)
+        // We want to startForeground so that the system is less likely to kill our service under memory pressure.
+        startForeground(
+            downloadJobs[download.id]?.foregroundServiceId ?: 0,
+            ongoingDownloadNotification
+        )
     }
 
-    private suspend fun performDownload(download: DownloadState) = withContext(IO) {
-        val headers = listOf(
-            CONTENT_TYPE to download.contentType,
-            CONTENT_LENGTH to download.contentLength?.toString(),
-            REFERRER to download.referrerUrl
-        ).mapNotNull { (name, value) ->
-            if (value.isNullOrBlank()) null else Header(name, value)
-        }.toMutableHeaders()
+    internal fun performDownload(download: DownloadState) {
+        val isResumingDownload = downloadJobs[download.id]?.currentBytesCopied ?: 0L > 0L
```

Comment:
Let's turn this expression into a function because we also need it below to set the header e.g. `getCurrentBytesCopied`.

Reply:
Yeah, should have marked it as `nit` maybe. It's mostly to simply the expression for the header value: `headers.append(RANGE,"bytes=${downloadJobs[download.id]?.currentBytesCopied}-")`